src/set_relation/set_relation.cc

```diff
@@ -2611,6 +2874,58 @@ void Relation::normalize() {
     // FIXME: will need more ... See SparseConstraints::normalize()
 }
 
+/* Mahdi Soltan Mohammadi
+   Is tuple variable tvar argument to an UFS? 
+*/
+bool SparseConstraints::isUFSArg(int tvar)
+{
+  bool is = false;
+
+    for (std::list<Conjunction*>::const_iterator i=mConjunctions.begin();
+                i != mConjunctions.end(); i++) {
+
+    Conjunction* conjunction = dynamic_cast<Conjunction*>((*i));
+    is = (conjunction->isUFSArg(tvar));
+    if ( is )  return is;
+}
+
+  return is;
+}
+
+
+/* Mahdi Soltan Mohammadi
+     Projects out tuple varrable No. tvar
+     tvar is calculated based on total ariety (in+out) starting from 0.
+     Consequently, to project out jp from R: tvar = 5
+     R = {[i,j,k] -> [ip,jp,kp] : ...}
+*/
+bool Relation::project_out(int tvar)
+{
+
+    if (isUFSArg(tvar)){
+      return false;
+    }
+
+    int ia = mInArity;
```

Comment:
This logic does need to be in Relation::project_out, but then the call to isUFSArg and calling project_out on the Constraints can be shared in SparseConstraints::project_out.